### DISCUSSION TITLE
Dragging and zooming in Plot.geo

### DISCUSSION THREAD

**Discussion by espinielli** [observablehq/plot#2258] [Q&A] [answered]
I used to zoom and drag maps using _base d3_…I wonder whether something like [d3-versor-dragging](https://observablehq.com/@d3/versor-dragging) is applicable to Plot.geo…or something else, BTW.

Thanks for any feedbacks

**Comment by Fil** [comment, marked as answer]
Hi Enrico,

here’s a port of the D3 notebook for Plot: [Plot: Versor Zooming / Observable | Observable](https://observablehq.com/@observablehq/plot-versor-zooming)

I fixed several issues since yesterday's version:
- the behavior is now local to the mark
- a pointer-events: all style on the plot ensures that it works on mobile
- the behavior also covers a dot mark

You can disable the zooming part by specifying a zoomExtent `[1, 1]`.

The centroid transform is not great for labels, because when a shape extends on both sides of the map, its centroid might be in the middle. See https://github.com/observablehq/plot/pull/2098 for a better alternative.

**Reply by espinielli** [reply]
Great Fil!
I did apply you code of yesterday to [World FIRs with zoom and rotate](https://observablehq.com/@espinielli/world-firs-with-zoom-and-rotate) and for positioning the labels I did finally use `Plot.geoCentroid()` which seems ok.
I will apply next your new version and report back.

**Reply by espinielli** [reply]
Done, but isn't `renderAlternative()` specialized for the two land geo's?
In my case I have other layers on top, i.e. FIRs.
(sorry if the question is stupid...)

**Reply by Fil** [reply]
Yes, for performance. because we want to render the less detailed geometries when zooming, and the more detailed geometries at the end. In your case all the FIRS are simple shapes so you would use the normal render function.

**Reply by Fil** [reply]
you'll need to compute land110m as a geoJSON too, not a topojson.

```
land110m = FileAttachment("land-110m.json")
  .json()
  .then((world) => topojson.feature(world, world.objects.land))
```

**Reply by espinielli** [reply]
great, that's what I did (without much understanding ;-) )
Thanks!

**Reply by Fil** [reply]
also the Bertin projection is not really meant to be rotated!

**Reply by espinielli** [reply]
I know but I like to start with that and then zoom and rotate...once users have tools they can harm themselves ;-)

**Reply by espinielli** [reply]
Strangely enough, in [World FIRs with zoom and rotate](https://observablehq.com/@espinielli/world-firs-with-zoom-and-rotate) if I rotate, say, 180 degrees I lose the `Plot.text` labels and have to re-run the chart to get them repainted...

**Reply by Fil** [reply]
Good observation. That's due to projection clipping. Points that are clipped initially (because they are on the other side of the globe) always are invisible, and conversely points that are visible initially always are visible (even if you rotate them out). We can use https://observablehq.com/@d3/testing-projection-visibility to solve this (I've updated the notebook accordingly).

**Reply by espinielli** [reply]
you are stratospheric!
Thanks!
